router: make full map_callrw with split args

mrForza opened the pull request:
This patch introduces a new way of `map_callrw` execution by which we can pass some arguments to all storages and split buckets' arguments to those storages that have at least one bucket of `bucket_ids`. To achieve this we introduce a new string option - `mode` to `map_callrw` api.

@TarantoolBot document

Title: vshard: `mode` option for `router.map_callrw()`

This string option regulates on which storages the user function will be executed via `map_callrw`. Possible values:
1) `full` - the user function will be executed on all storages in
   cluster.
2) `partial` - the user function will be executed on certain sotrages on
   which buckets from `bucket_ids` were found.

After that changes `map_callrw` works in 4 different scenarios depending on `mode` and `bucket_ids` options:
1) `<mode = 'full', bucket_ids = nil>` - user function will be executed
   with `args` on all storages.
2) `<mode = 'full', bucket_ids = {[bid_1] = {b_arg_1}, ...}>` - storages
   that have at least one bucket of `bucket_ids` will execute user
   function with `args` and additional buckets' arguments. Other storages
   will execute user function only with `args`.
3) `<mode = 'partial', bucket_ids = {bid_1, ...}>` - user function will be
   executed with `args` on storages that have at least one bucket of
   `bucket_ids`.
4) `<mode = 'partial', bucket_id = {[bid_1] = {b_arg_1}, ...}>` - the same
   as the 3rd scenario but buckets' arguments (b_arg_1, ..., b_arg_n) will
   be added as additional arguments to user function.

Also now `map_callrw` ends with error in cases of `<mode = 'full', bucket_ids = {1, 2, ...}>` and `<mode = 'partial', bucket_ids = nil>`.

Closes tarantool/vshard#559

Gerold103 commented on vshard/router/init.lua:
```diff
@@ -978,15 +980,19 @@ local function replicasets_map_reduce(replicasets, rid, func, args,
+        if grouped_args_added then
             table.remove(args)
+            grouped_args_added = false
         end
```
You won't need this flag I think if you would change it to something as:
```Lua
for ... do
    local rs_args = grouped_args and grouped_args[id] or nil
    if rs_args then table.insert(args, grouped_args[id]) end
    -- do the call
    if rs_args then table.remove(args) end
end
```
The code style is a bit relaxed, but you get the idea.

Gerold103 commented on vshard/router/init.lua:
```diff
@@ -1083,14 +1089,46 @@ local function router_group_map_callrw_args(router, bucket_ids, bucket_args)
+local function router_set_map_callrw_mode(opts_mode, bucket_ids)
+    if opts_mode == nil then
+        local mode = bucket_ids and MAP_CALLRW_PARTIAL or MAP_CALLRW_FULL
+        return mode, nil
```
You don't need to return trailing `nil`s in Lua. Can be simplified to `return bucket_ids and MAP_CALLRW_PARTIAL or MAP_CALLRW_FULL`.

Gerold103 commented on vshard/router/init.lua:
```diff
@@ -1137,7 +1177,12 @@ end
     if opts then
+        mode, err = router_set_map_callrw_mode(opts.mode, opts.bucket_ids)
+        if err then
+            return nil, err, nil
```
Same here. No need for a trailing `nil`.

Gerold103 commented on vshard/router/init.lua:
```diff
@@ -1137,7 +1177,12 @@ end
+            return nil, err, nil
+        end
         timeout = opts.timeout or consts.CALL_TIMEOUT_MIN
         do_return_raw = opts.return_raw
```
Here you assign `mode` and then overwrite it when `opts` are provided. Lets not do an overwrite. And instead simply declare the empty mode variable, and with given `opts` do what you do now, and in the `else` branch assign it to the "full" default. Then there won't be an overwrite, which in turn might save us a tiny bit of cpu.

Gerold103 commented on vshard/router/init.lua:
```diff
@@ -1083,14 +1089,46 @@ local function router_group_map_callrw_args(router, bucket_ids, bucket_args)
+--       bucket_ids.
+--    4) <mode = 'partial', bucket_id = {[bid_1] = {b_arg_1}, ...}> - the same
+--       as the 3rd scenario but buckets' arguments (b_arg_1, ..., b_arg_n)
+--       will be added as additional arguments to user function.
```
This here and in the commit message might be slightly misleading. It looks like the user should expect the bucket arguments to be passed as positional arguments like `func(all_args, bucket1_arg, bucket2_arg, ...)`. But instead it is more like `func(all_args, bucket_args_map)`. I suggest to clarify this.

Gerold103 reviewed (commented):
Thanks 🔥! I sure hope it will help the customers 😁.

Gerold103 commented:
The commit message doesn't mention the ticket. You need to write `Closes #559` right before the TarantoolBot doc request.

Serpentian commented on vshard/router/init.lua:
```diff
@@ -979,15 +981,12 @@ local function replicasets_map_reduce(replicasets, rid, func, args,
+        local rs_args = grouped_args and grouped_args[id] or nil
+        -- It's cheaper to push and then pop, rather then deepcopy
+        -- arguments table for every call.
+        if rs_args then table.insert(args, rs_args) end
```
Nit: let's preserve the old style with proper indentation of the conditions. It's easier to look at the code and minimizes the diff

mrForza replied:
fixed

Serpentian commented on vshard/router/init.lua:
```diff
@@ -1136,33 +1171,37 @@ end
             router_ref_storage_all(router, timeout)
     end
     if timeout then
+        if plain_bucket_ids then
```
This won't work unfortunately. Previously the `router_group_map_callrw_args` was invoked right after `router_ref_storage_by_buckets`, now it can be invoked also after `router_ref_storage_all`. The first one finds the buckets in the cluster, updates the router map cache and only because of that we could use the `router_group_map_callrw_args` function. But the `router_ref_storage_all` doesn't do that.

Currently, the newly added mode (mode = 'full' + split args) doesn't work, when the cache on a router is broken, the arguments will be sent to the wrong instances. Let's won't forget to add the test case for that

Serpentian reviewed (commented):
Thank you for the patch! It looks great, but I'm afraid, it's a little bit broken now, let's fix and test the new mode with broken cache on the router

mrForza commented on test/router-luatest/map_callrw_test.lua:
```diff
@@ -685,3 +685,116 @@ g.test_map_callrw_with_cdata_bucket_id = function(cg)
         ilt.assert_not(err)
```
fixed

Serpentian commented on test/router-luatest/map_callrw_test.lua:
```diff
@@ -685,3 +685,164 @@ g.test_map_callrw_with_cdata_bucket_id = function(cg)
+    -- We override 'do_map' function on storages in order to check that
+    -- default arguments and bucket arguments were successfully passed into
+    -- destination storages according to mode and bucket_ids options.
+    vtest.cluster_exec_each_master(cg, function()
```
It's not guaranteed, that the tests will run in order, you defined: for example you can ran a single test case with `test-run`, and it won't work in that case, since the `do_map` is overridden in another test case only.

We're supposed to reset the cluster at the same state, as it was before the test. So, we have to overwrite the `do_map` function at the start of every test and drop it at the end of each one.

mrForza replied:
fixed

Serpentian reviewed (commented):
Thank you for the patchset. Very good! Seems like the last comments from me

Serpentian reviewed (approved):
Thank you a lot! Clean work, IMHO. Hope, it'll help TDG make requests safer

mrForza commented on vshard/router/init.lua:
```diff
@@ -1132,40 +1179,48 @@ end
+        if not timeout then
+            goto fail
         end
     end
```
No, in full map_callrw with split args we must invoke `router_ref_storage_by_buckets`, because without using this function the `router_group_map_callrw_args` will build incorrect table of form `{rs_id = {b_id = {b_args}}}` as the router's cache may be broken. `router_ref_storage_by_buckets` finds the buckets in the cluster, updates the router's cache - `router.route_map` and refs storages which relate to given buckets ids whereas the `router_ref_storage_all` doesn't do that.  See [Nikita's comment](https://github.com/tarantool/vshard/pull/628#discussion_r2643341859) related to this problem.

mrForza committed:
router: refactoring of router_ref_storage_* functions

Before this patch the main `map_callrw` ref functions such as
`router_ref_storage_all` and `router_ref_storage_by_buckets` were
enormous (71 and 108 lines of code). Also these functions have a large
number of similar functional code blocks such as "sending refs",
"collecting refs" e.t.c. Since in tarantool/vshard#559 patch we will
extend the logic of full map_callrw making it able to work with split
args, the `router_ref_storage_all` can double in size. It can lead to
degradation of our codebase due to less readability.

To fix it we firstly determine general and repeated code blocks in ref
functions:
1) `ref-prepare`: groups buckets by replicasets with router's cache,
   builds a table of "target" replicasets and waits necessary masters.
2) `ref-send`: sends refs to the remote storage asynchronously and builds
   a table of future objects for the next processing.
3) `ref-collect`: waits until future objects are ready in order to
   extract payload from it (responses of storages' functions).
4) `ref-process`: a custom logic for `full` or `partial` map_callrw
   modes which describes how we should process results from future
   objects.

After defining the main stages of ref map_callrw functions we should
unify them so that we can use them in both `router_ref_storage_all` and
`router_ref_storage_by_buckets`.

Needed for tarantool/vshard#559

NO_TEST=refactoring
NO_DOC=refactoring

mrForza committed:
router: refactoring of ref_ids' logic in router_map_callrw

This patch takes initialization of `rid` out to `router_map_callrw` and
passes this variable to ref-functions. It is needed for future features
tidiness, for example - `make full map_callrw with split args` in which
the logic of `router_map_callrw` becomes more complex.

Needed for #559

NO_DOC=refactoring
NO_TEST=refactoring

mrForza committed:
test: refactoring of router/reload_test

Before this patch the `router-luatest/reload_test` checked router's
services only with old routers. However in future patch (gh-214) we
need to check map_callrw with old storages.

In order to make it able we:
1) change `vtest.cluster_new` so that we can pass server_config with
   certain ENV (LUA_PATH) variable into it. It can help us to create
   a new cluster on old version of vshard.
2) change `reload_router` to more general `reload_server` in order to
   unify the process of servers (router / storage) upgrade in
   `router-luatest/reload_test`.
3) unify the process of cluster creation -
   `create_cluster_on_specific_version` and the process of getting
   server's config with new ENV (LUA_PATH) variable -
   `get_config_for_specific_vshard_version`.

Needed for tarantool/vshard#214

NO_DOC=refactoring
NO_TEST=refactoring

mrForza committed:
router: make full map_callrw with split args

This patch introduces a new way of `map_callrw` execution by which we can
pass some arguments to all storages and split buckets' arguments to those
storages that have at least one bucket of `bucket_ids`. To achieve this we
introduce a new string option - `mode` to `map_callrw` api.

Also we change the logic of `router_ref_storage_all` ref function. Firstly
we ref all storages and get back an amount of "moved" buckets according
to the previously built router's cache. Then if there are no "moved"
buckets we accumulate and check total amount of buckets on all storages
and finish map_callrw ref stage. Otherwise, if there are some "moved"
buckets we perform the second network hop by checking on which replicasets
do the remaining "moved" buckets reside on.

Closes tarantool#559

@TarantoolBot document

Title: vshard: `mode` option for `router.map_callrw()`

This string option regulates on which storages the user function will be
executed via `map_callrw`. Possible values:
1) mode = 'partial'. In this mode user function will be executed on
   storages that have at least one bucket of 'bucket_ids'. The
   'bucket_ids' option can be presented in two ways: like a numeric array
   of buckets' ids or like a map of buckets' arguments. In first one user
   function will only receive args, in second one it will additionally
   receive buckets' arguments.
2) mode = 'full'. In this mode user function will be executed with args on
   all storages in cluster. Also we can pass 'bucket_ids' only like a map
   of bucket's arguments - `bucket_ids = {id = {args}, id = {args}, ...}`.
   In this way the user function will additionally receive buckets'
   arguments on those storages that have at least one bucket of
   `bucket_ids`.

If we didn't specify the 'mode' option, then it is set based on
'bucket_ids' option for backward compatibility. If 'bucket_ids' is
presented, the mode will be 'partial' otherwise 'full'.

Also now `map_callrw` ends with error in cases when we
1) pass `mode = 'full'` and a numeric array of buckets' ids -
   `bucket_ids = {id, id, ...}`
2) pass `mode = 'partial'` and `bucket_ids = nil`.

mrForza committed:
storage: refactoring of bucket_get_moved

In this patch we change `allstatus.GARBAGE/SENT` on `BACTIVE/BSENT` to
not repeat the code.

Needed for tarantool/vshard#214

NO_DOC=refactoring
NO_TEST=refactoring
